src/Microsoft.VisualStudio.Extensibilty.Testing.SourceGenerator/TestServicesSourceGenerator.cs

```diff
@@ -0,0 +1,1014 @@
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft;
+    using Microsoft.VisualStudio;
+    using Microsoft.VisualStudio.Shell;
+    using Microsoft.VisualStudio.Shell.Interop;
+    using Microsoft.VisualStudio.Threading;
+
+    [TestService]
+    internal partial class ShellInProcess
+    {
+        public new Task<TInterface> GetRequiredGlobalServiceAsync<TService, TInterface>(CancellationToken cancellationToken)
+            where TService : class
+            where TInterface : class
+        {
+            return base.GetRequiredGlobalServiceAsync<TService, TInterface>(cancellationToken);
+        }
+
+        public new Task<TService> GetComponentModelServiceAsync<TService>(CancellationToken cancellationToken)
+            where TService : class
+        {
+            return base.GetComponentModelServiceAsync<TService>(cancellationToken);
+        }
+
+        public async Task<string> GetActiveWindowCaptionAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var monitorSelection = await GetRequiredGlobalServiceAsync<SVsShellMonitorSelection, IVsMonitorSelection>(cancellationToken);
+            ErrorHandler.ThrowOnFailure(monitorSelection.GetCurrentElementValue((uint)VSConstants.VSSELELEMID.SEID_WindowFrame, out var windowFrameObj));
+            var windowFrame = (IVsWindowFrame)windowFrameObj;
+
+            ErrorHandler.ThrowOnFailure(windowFrame.GetProperty((int)__VSFPROPID.VSFPROPID_Caption, out var captionObj));
+            return $""{captionObj}"";
+        }
+
+        public async Task<Version> GetVersionAsync(CancellationToken cancellationToken)
+        {
+            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var shell = await GetRequiredGlobalServiceAsync<SVsShell, IVsShell>(cancellationToken);
+            shell.GetProperty((int)__VSSPROPID5.VSSPROPID_ReleaseVersion, out var versionProperty);
+
+            var fullVersion = versionProperty?.ToString() ?? string.Empty;
+            var firstSpace = fullVersion.IndexOf(' ');
+            if (firstSpace >= 0)
+            {
+                // e.g. ""17.1.31907.60 MAIN""
+                fullVersion = fullVersion.Substring(0, firstSpace);
+            }
+
+            if (Version.TryParse(fullVersion, out var version))
+            {
+                return version;
+            }
+
+            throw new NotSupportedException($""Unexpected version format: {versionProperty}"");
+        }
+    }
+}
+";
+
+        private const string SolutionExplorerInProcessSource = @"// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for more information.
```

Comment:
@sharwell So I get that some of the generated files are checking for specific optional APIs, but for the ones that are static, is this just to work around the PIA changes between Dev16 and Dev17? Is there any reason we can't just use NuGet content files for these instead? This would make this far easier to edit and maintain.

Reply:
> Is there any reason we can't just use NuGet content files for these instead?

* I find source generators preferable to NuGet content files. 
* It keeps everything in one place.
* Source generator test APIs don't understand NuGet content packages.